node/src/reactor/participating.rs

```diff
@@ -682,28 +682,80 @@ impl reactor::Reactor for Reactor {
                                 }
                             };
 
-                            match self
-                                .storage
-                                .handle_legacy_direct_deploy_request(deploy_hash)
-                            {
-                                // This functionality was moved out of the storage component and
-                                // should be refactored ASAP.
-                                Some(deploy) => {
-                                    match Message::new_get_response(&deploy) {
-                                        Ok(message) => {
-                                            return effect_builder
-                                                .send_message(sender, message)
-                                                .ignore();
-                                        }
-                                        Err(error) => {
-                                            error!("failed to create get-response: {}", error);
-                                            return Effects::new();
-                                        }
-                                    };
+                            // Depending on the runtime configuration, fetch as a shared or owned
+                            // object.
+                            if self.storage.mem_duplication_enabled() {
+                                let outcome =
+                                    self.storage.deploy_cache.get(&deploy_hash).map(Ok).or_else(
```

Comment:
Couldn't the caching be handled entirely inside `Storage` rather than requiring explicit cache read/writes here?

Reply:
That's a damn good catch. Moved everything inside a separate storage component method.